File: python/LinkedListCycleII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        record_nodes = set()
        cur = head
        while cur:
            if cur in record_nodes:
                return cur
            record_nodes.add(cur)
            cur = cur.next
        return None

File: java/LinkedListCycleII.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        HashSet<ListNode> record_nodes = new HashSet();
        ListNode cur = head;
        while(cur != null){
            if(record_nodes.contains(cur))
                return cur;
            record_nodes.add(cur);
            cur = cur.next;
        }
        return null;
    }
}
